File: strategy/marvin/agent/attacker/penaltyshootout.lua
local Base = require "agent/base/behavior"
local PenaltyShootout = Class("Agent.Attacker.PenaltyShootout", Base)

local Referee = require "../base/referee"
local World = require "../base/world"
local G = World.Geometry

local Goal = require "observer/goal"
local Robot = require "observer/robot"
local MoveToStaticBall = require "task/attacker/movetostaticball"
local ShootGoal = require "task/attacker/shootgoal"
local StopAttack = require "task/attacker/stopattack"
local MoveToBall = require "task/attacker/movetoball"
local Dribble = require "task/attacker/dribble"
local Pass = require "task/attacker/penaltyshootout"
local Chip = require "task/attacker/penaltychip"
local Field = require "../base/field"

local vis = require "../base/vis"
local debug = require "../base/debug"


local DISTANCE_TO_DEFENSE_AREA = 0.6 -- the furthest we'll go before we shoot
local MIN_RELATIVE_SECTOR_SIZE = 1/3
local DRIBBLING_DISTANCE = 0.075 -- Ball and Robot must be at least this far apart to reset dribbling


function PenaltyShootout:_stop()
	self._penaltyStartTime = nil
	self._contactPoint = nil
	self._shootGoalFlag = false
	self._forceDesperate = false
	self._changeContact = false
	self._baseDribblePos = Vector(0, G.FieldHeightHalf)
	self._addPos = Vector(0, 0)
	self._state = nil
	self._futureKeeper = {pos = World.Geometry.OpponentGoal, speed = Vector(0,0.1), radius = 0.09}
	self._lastKeeper = {pos = World.Geometry.OpponentGoal, speed = Vector(0,0.1), radius = 0.09}
	self._ball = {pos = World.Ball.pos, speed = World.Ball.speed, radius = World.Ball.radius, time = World.Time}
	self._dribblePos = nil
end

function PenaltyShootout:check()
	local mainAttacker = self._inbox.mainAttacker().trainer == self._robot
	local isPenalty = World.RefereeState == "PenaltyOffensivePrepare" or World.RefereeState == "PenaltyOffensive"
	local isShootout = World.GameStage == "PenaltyShootout"
	-- log("")
	-- log("check")
	-- log("mainAttacker: "..tostring(mainAttacker))
	-- log("isPenalty: "..tostring(isPenalty))
	-- log("isShootout: "..tostring(isShootout))
	-- log("onGoing: "..tostring(self:_checkPenaltyOngoing()))
	return mainAttacker and isShootout and (isPenalty or self:_checkPenaltyOngoing())
end

function PenaltyShootout:_checkPenaltyOngoing()
	return self._penaltyStartTime and World.Time - self._penaltyStartTime < 15 and not Referee.isStopState()
end

function PenaltyShootout:_updateDribbling()
	-- log("update")
	if not self._contactPoint and Robot.hadBall(self._robot, 0) then
		-- log("1")
		self._contactPoint = self._robot.pos
		self._changeContact = true
	elseif self._contactPoint and World.Ball.pos:distanceTo(self._robot.pos) > DRIBBLING_DISTANCE + self._robot.radius then
		-- log("2")
		self._contactPoint = nil
		self._changeContact = true
	else
		self._changeContact = false
	end
end

function PenaltyShootout:_updateShootGoal()
	if World.OpponentKeeper and World.OpponentKeeper.pos then
		self._futureKeeper = {pos = World.OpponentKeeper.pos, radius = World.OpponentKeeper.radius}
	end
	local lastContact = self._contactPoint
	local addDistance = lastContact and math.max(0, lastContact:distanceTo(World.Ball.pos) - 0.5)*3 or 0.2
		self._futureKeeper.pos = self._futureKeeper.pos + self._lastKeeper.speed * 0.4
	if self._state == "pass" then
		self._futureKeeper.pos = self._futureKeeper.pos + (self._robot.pos - self._futureKeeper.pos):setLength(self._robot.speed:length()/3 + addDistance/4)
	end

	debug.push("Shootgoal Criterias")
	if self._penaltyStartTime then
		local timeSinceStart = World.Time - self._penaltyStartTime
		local criteriaTime = timeSinceStart > 8
		debug.push("Time Criteria (8s)", criteriaTime)
		debug.set("timeSinceStart", timeSinceStart)
		debug.pop()

		local ballPosY = World.Ball.pos.y
		local distanceToGoalLine = (World.RULEVERSION == "2017" and G.DefenseRadius or G.DefenseHeight) + DISTANCE_TO_DEFENSE_AREA
		local criticalMark = G.FieldHeightHalf - distanceToGoalLine
		local criteriaPos = ballPosY + addDistance > criticalMark
		debug.push("Position Criteria", criteriaPos)
		debug.set("ballPosY", ballPosY)
		debug.set("addDistance", addDistance)
		debug.set("DistanceToGoalLine", distanceToGoalLine)
		debug.set("CriticalMark", criticalMark)
		debug.pop()
		vis.addCircle("a/a/penaltyshootout: futureKeeper", self._futureKeeper.pos, 0.1, vis.colors.green, false)

		local sector = Goal.largestFreeSector(World.Ball.pos, {self._futureKeeper}, true)
		local width = sector and math.abs(sector[1] - sector[2]) or 0
		local angle = 2 * math.atan((G.GoalWidth / 2) / (G.FieldHeightHalf - self._robot.pos.y))
		local criteriaAngle = width < angle * MIN_RELATIVE_SECTOR_SIZE
		debug.push("Angle Criteria", criteriaAngle)
		debug.set("width", width*180/math.pi)
		debug.set("minRelativeSectorSize", MIN_RELATIVE_SECTOR_SIZE)
		debug.set("maxAngleForPosition(in deg)", 180/math.pi * angle)
		debug.pop()
		if self._shootGoalFlag or criteriaTime or criteriaTime or criteriaAngle then
			self._shootGoalFlag = true
		end
	else
		debug.push("Time Criteria (8s)")
		debug.set("time", "not set yet")
		debug.pop()
	end
	debug.pop()
end

--- Calculates the effective distance between ball and dribbler
-- find an ellipsis with the left and right dribbler edge points as focal points
-- dist is the length of the semi-minor axis
-- @param robot robot - the robot to calculate
-- @param ballPos vector - position of the ball
local function ellipticDistance(robot, ballPos)
	local dribblerPos = robot.pos + Vector.fromAngle(robot.dir):scaleLength(robot.shootRadius)
	local dribblerWidthHalf = Vector.fromAngle(robot.dir - math.pi/2):scaleLength(robot.dribblerWidth/2)
	local leftDribblerEdge = dribblerPos + dribblerWidthHalf
	local rightDribblerEdge = dribblerPos - dribblerWidthHalf
	return 0.5*math.sqrt((leftDribblerEdge:distanceTo(ballPos) + rightDribblerEdge:distanceTo(ballPos))^2 - robot.dribblerWidth*robot.dribblerWidth)
end
function PenaltyShootout:_updateBall()
	if self._ball == nil then
		self._ball = {pos = World.Ball.pos, speed = World.Ball.speed, radius = World.Ball.radius, time = World.Time}
	end
	local minDistance = self._robot.radius + self._ball.radius - 0.01
	local ballPos = self._ball.pos
	local robotPos = self._robot.pos
	local t_now = World.Time
	if World.Ball:isPositionValid() then
		if robotPos:distanceToSq(World.Ball.pos) < minDistance * minDistance then
			if ellipticDistance(self._robot, self._ball.pos) < self._ball.radius + 0.01 then
				local dribblerPos = robotPos + Vector.fromAngle(self._robot.dir):scaleLength(self._robot.shootRadius)
				local a = 0.9
				ballPos = ballPos * a + dribblerPos * (1-a)
			end
			local alpha = 0.8
			local relSpeed = self._ball.speed - self._robot.speed
			local reflect = relSpeed + (ballPos - robotPos) * (ballPos - robotPos):dot(relSpeed)/robotPos:distanceToSq(ballPos) * -2
			self._ball.speed = self._robot.speed * alpha + reflect * (1-alpha)
			self._ball.pos = self._ball.speed * (t_now - self._ball.time) + ballPos
			self._ball.time = t_now
		else
			self._ball.pos = World.Ball.pos
			self._ball.speed = World.Ball.speed
			self._ball.time = t_now
		end
	else
		self._ball.pos = (t_now - self._ball.Time)*self._ball.speed + ballPos
		self._ball.time = t_now
	end
	vis.addCircle("a/a/penalty: ball",self._ball.pos, 0.05, vis.colors.orchid, true)
end

function PenaltyShootout:_updateTask()
	if World.OpponentKeeper and World.OpponentKeeper.pos then
		self._lastKeeper = World.OpponentKeeper
	end
	self:_updateBall()
	self:_updateDribbling()
	self:_updateShootGoal()
	debug.set("ShootGoalFlag", self._shootGoalFlag)
	local keeperPos = self._lastKeeper.pos
	local lastContact = self._contactPoint
	local robotPos = self._robot.pos
	local freeway = self._state == "pass" and 0.1 or 0
	local distanceToContact
	if lastContact then
		vis.addCircle("1test", lastContact, 1, vis.colors.green, false)
		distanceToContact = lastContact:distanceTo(self._ball.pos)
	else
		distanceToContact = 0
	end

	if World.RefereeState == "PenaltyOffensive" and not self._penaltyStartTime then
		-- log("Start Time set")
		self._penaltyStartTime = World.Time
	end
	local chipMode = Chip.check(self._ball, self._robot)
	if World.RefereeState == "PenaltyOffensivePrepare" then
		return MoveToStaticBall, {math.pi / 2, 0.1}
	elseif chipMode then
		return Chip, {self._ball, chipMode}
	elseif self._shootGoalFlag then
		return ShootGoal--, nil, true
	elseif distanceToContact > 1 + 0.3 then
		return ShootGoal
	elseif not lastContact or robotPos:distanceTo(self._ball.pos) > self._robot.radius + World.Ball.radius + freeway then
		return MoveToBall, {0.01}
	elseif distanceToContact > 1 - 0.05 then
		return StopAttack
	elseif distanceToContact > 1 - 0.3 then
		self._state = "pass"
		return Pass, {self._ball}
	elseif distanceToContact > 1 - 0.35 then
		return MoveToBall, {0.00}
	else

		self._state = "dribble"
		-- self._robot:setDribblerSpeed(0.5)
		-- return MoveToBall, {-0.1}, self._changeContact
		local rate = 0.02 * robotPos:distanceTo(keeperPos)
		-- if lastContact:distanceTo(World.Ball.pos) < 1 - 0.2 then
			if keeperPos.x < 0 then
				self._addPos.x = (self._addPos.x + rate) / (1+math.abs(robotPos.x - keeperPos.x))
			else
				self._addPos.x = (self._addPos.x - rate) / (1+math.abs(robotPos.x - keeperPos.x))
			end
		-- else
		-- 	self._addPos.x = 0
		-- end
		local dribblePoint = self._baseDribblePos + self._addPos
		local intersection = Field.intersectRayDefenseArea(dribblePoint, robotPos - dribblePoint, 0.2, false)
		if intersection then
			dribblePoint = intersection
		end
		if self._dribblePos == nil then
			self._dribblePos = dribblePoint
		else
			self._dribblePos.x = dribblePoint.x
			self._dribblePos.y = dribblePoint.y
		end
		return Dribble, {self._dribblePos}--, true
		-- return Dribble, {dribblePoint}--, true
	end
end

return PenaltyShootout

File: strategy/typescript/glados/agent/attacker/penaltyshootout.ts
import * as Const from "base/constants";
import * as debug from "base/debug";
import * as Field from "base/field";
import * as geom from "base/geom";
import * as Referee from "base/referee";
import { Robot } from "base/robot";
import { AbsTime, RelTime } from "base/timing";
import { Position, Speed, Vector } from "base/vector";
import * as vis from "base/vis";
import * as World from "base/world";

import { Behavior, TaskAssignment } from "glados/agent/base/behavior";
import { MessageType } from "glados/control/messaging";
import { ellipticDistance, getRealisticBallPos  } from "glados/observer/ball";
import * as Goal from "glados/observer/goal";
import * as Physics from "glados/observer/physics";
import * as ORobot from "glados/observer/robot";
import { Dribble } from "glados/task/attacker/dribble";
import { MoveToBall } from "glados/task/attacker/movetoball";
import { MoveToStaticBall } from "glados/task/attacker/movetostaticball";
import { PenaltyChip } from "glados/task/attacker/penaltychip";
// import { PenaltyShootout as Pass } from "glados/task/attacker/penaltyshootout";
import { PenaltyShootoutGoal as ShootGoal } from "glados/task/attacker/penaltyshootoutgoal";
import { StopAttack } from "glados/task/attacker/stopattack";
import { MoveToPos } from "glados/task/shared/movetopos";


let G = World.Geometry;

let DISTANCE_TO_DEFENSE_AREA = 0.6; // the furthest we'll go before we shoot
let DRIBBLING_DISTANCE = 0.035; // Ball and Robot must be at least this far apart to reset dribbling

function getKeeperTime(post: Position, turntime: RelTime, ball: {pos: Position, speed: Speed, radius: number, maxSpeed: number}, keeper: Robot) {
	let pos = ball.pos + ball.speed * turntime;
	let startspeed = Const.maxBallSpeed - Const.fastBallDeceleration * turntime;
	let shootdir = (post - pos).normalized();
	let startpos = pos - shootdir * ((Const.maxBallSpeed + startspeed) / 2 * turntime);

	let startBall = {pos : startpos, speed : shootdir * startspeed,
			radius : ball.radius, maxSpeed : Const.maxBallSpeed};
	if (ball.speed.lengthSq() > 3 * 3) {
		startBall = ball;
	}
	vis.addPath("a/a/penaltyshootout: startBall", [startpos, pos], vis.colors.green);
	vis.addPath("a/a/penaltyshootout: startBall", [pos, post], vis.colors.black);
	let time: number = Physics.robotTimeToBall(keeper, startBall, G.OpponentGoal,0);
	if (time !== Infinity) return 0;
	let dir = shootdir.perpendicular();
	let r = (keeper.pos - startpos).dot(dir);
	let f = (r < 0 ? -1 : 1);
	r = Math.max(f * r - Const.maxRobotRadius, 0.001);
	let v = keeper.speed.dot(dir) * f;
	let a = - keeper.acceleration.aSpeedupSMax;
	time = (-v + Math.sqrt(v * v - 4 * a * r)) / (2 * a);
	// }
	return time;
}

export class PenaltyShootout extends Behavior {
	private _penaltyStartTime: number | undefined = undefined;
	private _contactPoint: Position | undefined = undefined;
	private _shootGoalFlag: boolean = false;
	private _forceDesperate: boolean = false;
	private _changeContact: boolean = false;
	private _baseDribblePos: Position = new Vector(0, G.FieldHeightHalf);
	private _addPos: Position = new Vector(0, 0);
	private _state: string | undefined;
	private _futureKeeper: {pos: Position, speed: Speed, radius: number} = {pos: G.OpponentGoal, speed: new Vector(0,0.1), radius: 0.09};
	private _ball = {pos : getRealisticBallPos(), speed : World.Ball.speed, radius : World.Ball.radius, time : World.Time, maxSpeed: 0, quality : 1};
	private _dribblePos: Position | undefined = undefined;
	private _shootPos: Position = G.OpponentGoal;

	_stop() {

		this._penaltyStartTime = undefined;
		this._contactPoint = undefined;
		this._shootGoalFlag = false;
		this._forceDesperate = false;
		this._changeContact = false;
		this._baseDribblePos = new Vector(0, G.FieldHeightHalf);
		this._addPos = new Vector(0, 0);
		this._state = undefined;
		this._futureKeeper = {pos: G.OpponentGoal, speed: new Vector(0,0.1), radius: 0.09};
		this._ball = {pos : getRealisticBallPos(), speed : World.Ball.speed, radius : World.Ball.radius, time : World.Time, maxSpeed: 0, quality : 1};
		this._dribblePos = undefined;
	}

	public check(): boolean {
		let mainAttacker = this._messaging.receiveTrainer(MessageType.mainAttacker) === this._robot;
		let isPenalty = World.RefereeState === "PenaltyOffensivePrepare" || World.RefereeState === "PenaltyOffensive";
		let isShootout = World.GameStage === "PenaltyShootout";
		return mainAttacker && isShootout && (isPenalty || this._checkPenaltyOngoing());
	}

	private _checkPenaltyOngoing(): boolean {
		return this._penaltyStartTime != undefined && World.Time - this._penaltyStartTime < 15 && !Referee.isStopState();
	}

	private _updateDribbling() {
		if (this._contactPoint == undefined && ORobot.hadBall(this._robot, 0)) {
			this._contactPoint = this._robot.pos;
			this._changeContact = true;
		} else if (this._contactPoint != undefined &&
				getRealisticBallPos().distanceTo(this._robot.pos) > DRIBBLING_DISTANCE + this._robot.radius + World.Ball.radius) {
			this._contactPoint = undefined;
			this._changeContact = true;
		} else {
			this._changeContact = false;
		}
	}

	private angleUncertanty: number = 3 * Math.PI / 180;
	private turnspeed: number = 1;

	private _checkFreeGoal() {
		if (World.OpponentKeeper === undefined) {
			return [0.5, 0, 0];
		}
		let keeper = World.OpponentKeeper;
		let robot = this._robot;
		let ball = this._ball;
		let toleftPost = (G.OpponentGoalLeft - robot.pos).rotated(-this.angleUncertanty);
		let torightPost = (G.OpponentGoalRight - robot.pos).rotated(this.angleUncertanty);
		let anglediffleft = geom.normalizeAngle(toleftPost.angle() - robot.dir);
		let anglediffright = geom.normalizeAngle(torightPost.angle() - robot.dir);
		let turntimeLeft = Math.abs(anglediffleft) / this.turnspeed;
		let turntimeRight = Math.abs(anglediffright) / this.turnspeed;


		let timeLeft = getKeeperTime(robot.pos + toleftPost, turntimeLeft, ball, keeper);
		let timeRight = getKeeperTime(robot.pos + torightPost, turntimeRight, ball, keeper);

		if (timeRight === 0 && timeRight === 0) {
			return [false, 0, 0];
		}
		let weightLeft = timeLeft / (turntimeLeft + 0.2);
		let weightRight = timeRight / (turntimeRight + 0.2);
		let weight = weightRight / (weightLeft + weightRight);
		if (Number.isNaN(weight)) {
			debug.set("._.", String(timeLeft));
			debug.set("D;", String(timeRight));
		}
		if (weight > 0.5) {
			this._shootPos = robot.pos + torightPost;
		} else {
			this._shootPos = robot.pos + toleftPost;
		}
		return [weight,timeRight,timeLeft];
	}
	private _updateShootGoal() {
		if (World.OpponentKeeper  &&  World.OpponentKeeper.pos) {
			this._futureKeeper = {pos : World.OpponentKeeper.pos, speed: World.OpponentKeeper.speed, radius : World.OpponentKeeper.radius};
		}
		let lastContact = this._contactPoint;
		let addDistance = lastContact ? Math.max(0, lastContact.distanceTo(getRealisticBallPos()) - 0.5) * 3 : 0.2;
			this._futureKeeper.pos = this._futureKeeper.pos + (World.OpponentKeeper || this._futureKeeper).speed * 0.4;
		if (this._state === "pass") {
			this._futureKeeper.pos = this._futureKeeper.pos + (this._robot.pos - this._futureKeeper.pos).withLength(this._robot.speed.length() / 3);
		}

		debug.push("Shootgoal Criterias");
		if (this._penaltyStartTime != undefined) {
			let timeSinceStart = World.Time - this._penaltyStartTime;
			let criteriaTime = timeSinceStart > 8;
			debug.push("Time Criteria (8s)", String(criteriaTime));
			debug.set("timeSinceStart", timeSinceStart);
			debug.pop();

			let ballPosY = getRealisticBallPos().y;
			let distanceToGoalLine = (World.RULEVERSION === "2017" ? G.DefenseRadius : G.DefenseHeight) + DISTANCE_TO_DEFENSE_AREA;
			let criticalMark = G.FieldHeightHalf - distanceToGoalLine;
			let criteriaPos = ballPosY + addDistance > criticalMark;
			debug.push("Position Criteria", String(criteriaPos));
			debug.set("ballPosY", ballPosY);
			debug.set("addDistance", addDistance);
			debug.set("DistanceToGoalLine", distanceToGoalLine);
			debug.set("CriticalMark", criticalMark);
			debug.pop();
			vis.addCircle("a/a/penaltyshootout: futureKeeper", this._futureKeeper.pos, 0.1, vis.colors.green, false);

			let [weight,tright,tleft] = this._checkFreeGoal();
			debug.push("Free Goal Criteria", String(weight));
			debug.set("TimeRight", String(tright));
			debug.set("TimeLeft", String(tleft));
			debug.pop();

			if (this._shootGoalFlag  ||  criteriaPos  ||  criteriaTime  ||  // criteriaAngle ||
					(weight !== false && ellipticDistance(this._robot, this._ball.pos) < this._ball.radius + 0.02)) {
				this._shootGoalFlag = true;
			}
		} else {
			debug.push("Time Criteria (8s)");
			debug.set("time", "not set yet");
			debug.pop();
		}
		debug.pop();
	}

	private _updateBall() {
		let ball = this._ball;
		let quality = World.Ball.detectionQuality - ball.quality;
		let robot = this._robot;
		let minDistance = robot.radius + ball.radius - 0.01;
		let ballPos = ball.pos;
		let robotPos = robot.pos;
		let t_now = World.Time;
		debug.set(".ballVisible", World.Ball.isPositionValid());
		debug.set(".ballQuality", String(World.Ball.detectionQuality));
		if (World.Ball.isPositionValid()) {
			if (robotPos.distanceToSq(getRealisticBallPos()) < minDistance * minDistance || quality < 0) {
				if (ellipticDistance(robot, ball.pos) < ball.radius + 0.01) {
					let dribblerPos = robotPos + Vector.fromPolar(robot.dir, robot.shootRadius + ball.radius);
					let a = 0.9;
					ballPos = (ballPos * a) + (dribblerPos * (1 - a));
					// ballPos = (ballPos * a);
				}
				let alpha = 0.8;
				let relSpeed = ball.speed - robot.speed;
				let reflect = relSpeed + (ballPos - robotPos) * (ballPos - robotPos).dot(relSpeed) / robotPos.distanceToSq(ballPos) * -2;
				ball.speed = robot.speed * alpha + reflect * (1 - alpha);
				ball.pos = ball.speed * (t_now - ball.time) + ballPos;
				ball.quality = World.Ball.detectionQuality * 0.5 + ball.quality * 0.5;
			} else {
				ball.pos = getRealisticBallPos();
				ball.speed = World.Ball.speed;
				ball.quality = World.Ball.detectionQuality;
			}
		} else {
			ball.pos = (t_now - ball.time) * ball.speed + ballPos;
		}
		ball.time = t_now;
		vis.addCircle("a/a/penalty: ball", ball.pos, 0.05, vis.colors.orchid, true);
	}

	_updateTask() : TaskAssignment<typeof ShootGoal> | TaskAssignment<typeof MoveToStaticBall> |
			TaskAssignment<typeof MoveToBall> | TaskAssignment<typeof StopAttack> |
			TaskAssignment<typeof PenaltyChip> | TaskAssignment<typeof Dribble> |
			TaskAssignment<typeof MoveToPos> {
		this._updateBall();
		this._updateDribbling();
		this._updateShootGoal();
		debug.set("ShootGoalFlag", this._shootGoalFlag);
		let lastContact = this._contactPoint;
		let robotPos = this._robot.pos;
		let freeway : number = this._state === "pass" ? 0.1 : 0;
		let distanceToContact;
		if (lastContact) {
			vis.addCircle("1test", lastContact, 1, vis.colors.green, false);
			distanceToContact = lastContact.distanceTo(this._ball.pos);
		} else {
			distanceToContact = 0;
		}

		if (World.RefereeState === "PenaltyOffensive" && this._penaltyStartTime == undefined) {
			this._penaltyStartTime = World.Time;
		}
		let chipMode = PenaltyChip.check(this._ball, this._robot);
		if (World.RefereeState === "PenaltyOffensivePrepare") {
			return [MoveToStaticBall, [Math.PI / 2, 0.1]];
		} else if (distanceToContact > 1 - 0.05) {
			// return [StopAttack];
			return [ MoveToPos, [World.Ball.pos - new Vector(0, 0.5), (new Vector(0, -1)).angle(),
				undefined, undefined, undefined, undefined, undefined, undefined, true]];
		} else if (chipMode !== false) {
			debug.set("chipMode", chipMode);
			return [PenaltyChip, [this._ball, chipMode]];
		} else if (!lastContact  ||  robotPos.distanceTo(this._ball.pos) > this._robot.radius + World.Ball.radius + freeway) {
			return [MoveToBall, [G.OpponentGoal - this._robot.pos]];
		} else if (this._shootGoalFlag) {
			return [ShootGoal,[this._shootPos, this._ball]];// , nil, true
		} else {

			this._state = "dribble";
			let keeperPos = (World.OpponentKeeper || this._robot).pos;
			let rate = 0.02 * robotPos.distanceTo(keeperPos);
			if (keeperPos.x < 0) {
				this._addPos.x = (this._addPos.x + rate) / (1 + Math.abs(robotPos.x - keeperPos.x));
			} else {
				this._addPos.x = (this._addPos.x - rate) / (1 + Math.abs(robotPos.x - keeperPos.x));
			}
			let dribblePoint = this._baseDribblePos + this._addPos;
			let intersection = Field.intersectRayDefenseArea(dribblePoint, robotPos - dribblePoint, 0.2, false)[0];
			if (intersection) {
				dribblePoint = intersection;
			}
			if (this._dribblePos == undefined) {
				this._dribblePos = dribblePoint;
			} else {
				this._dribblePos.x = dribblePoint.x;
				this._dribblePos.y = dribblePoint.y;
			}
			return [Dribble, [this._dribblePos]];
		}
	}
}
